src/content/learning/using-durable-objects.md

```diff
@@ -0,0 +1,216 @@
+
+There are four steps to creating a Durable Object:
+
+* __Writing the class__ that defines a Durable Object.
+* __Binding that class__ into a Worker.
+* __Instantiating a Durable Object__ from within a running Worker.
+* __Communicating with a Durable Object__ from a running Worker via the Fetch API.
+
+## Writing a class that defines a Durable Object
+
+To define a Durable Object, developers export an ordinary Javascript class.  Other languages will need a shim that translates their class definition to a Javascript class.
+
+The first variable passed to the class constructor contains state specific to the Durable Object, including methods for accessing storage.
+
+```js
+
+class DurableObjectExample {
+    constructor(state, environment){
+
+    }
+}
+
+```
+
+Workers communicate with a Durable Object via the fetch API.  Like any other Worker, a Durable Object listens for incoming Fetch events by registering an event handler.
+
+```js
+
+class DurableObjectExample {
+    constructor(state, environment){
+    }
+    
+    addEventListener("fetch", (event) => {
+        event.respondWith(handle(event.request));
+    });
+ 
+    async function handle(request) {
+        return new Response('Hello World');
+    }
+
+}
+
+```
+
+A Worker can pass information to a Durable Object via headers, the HTTP method, the Request body, or the Request URI.
+
+### Accessing Storage from a Durable Object
+
+Durable Objects gain access to a TODO: fill in[persistent storage API]() via the first parameter passed to the Durable Object constructor.  While access to a Durable Object is single-threaded, it's important to remember that there can still be race conditions across multiple requests.
+
+```js
+
+class DurableObjectExample {
+    constructor(state, environment){
+        this.state = state
+    }
+    
+    addEventListener("fetch", (event) => {
+        event.respondWith(handle(event.request));
+    });
+ 
+    async function handle(request) {
+        let ip = request.headers.get('CF-Connecting-IP');
+        let data = request.text();
+        let storagePromise = this.state.storage.set(ip, data);
+        await storagePromise;
+        return new Response(ip + ' stored ' + data);
+    }
+
+}
+
+```
+
+Single statement transactions are always transactional.  More complex use cases can wrap multiple storage statements in a transaction.  
+
+```js
+
+class DurableObjectExample {
+    constructor(state, environment){
+        this.state = state
+    }
+    
+    addEventListener("fetch", (event) => {
+        event.respondWith(handle(event.request));
+    });
+ 
+    async function handle(request) {
+        // TODO: txn example
+    }
+
+}
+
+```
+
+Transactions can fail if they conflict with a concurrent transaction from the same Durable Object.  Transactions are transparently and automatically retried once before returning a failure.
```

Comment:
I'd insert an extra little clause here, changing "retried once before" to "retried once by rerunning the provided function before"

Reply:
This could also be a place to add in that this is because all transactions are run at serializable isolation, making them strongly consistent. Totally optional though